### ISSUE TITLE
DetailsList A11y Pass

### ISSUE THREAD

**Issue by TestAccessibilityGIT** [OfficeDev/office-ui-fabric-react#6646]
Tracking all of the issues brought up with the 10/09 accessibility pass on DetailsList  control
Details on individual bugs can be found on [SharePoint](https://microsoft-my.sharepoint.com/:x:/r/personal/v-amaver_microsoft_com/_layouts/15/Doc.aspx?sourcedoc=%7B9694606d-47a5-4027-8aba-e9569be6adad%7D&action=edit&activeCell='Bugs'!D8&wdInitialSession=07a2e8c8-5abd-4ed7-8e4e-ebe1702ba085&wdRldC=1)

<s>- Bug 1: [Accessibility][Programmatic access]: Buttons must have discernible text.</s>
- [x] Bug 2: [Accessibility][Describing content, controls, or links]: Images must have alternate text.
- [x] Bug 3: [Accessibility][Programmatic Access]: 'Column sorting' Information does not announce by screen reader. [Duplicated by #6011, #10168]
  - 8/16/19: @JasonGore While investigating #10168, I've confirmed this is still an open and active issue against Narrator.
  - 10/29/18 @KevinTCoughlin: This is true, though the column header has `aria-sort="descending | ascending"` set. Also the official ARIA example with `aria-sort` does not announce by screen reader either. See https://www.w3.org/TR/wai-aria-practices/examples/grid/dataGrids.html.
- [x] Bug 4: [Accessibility][Programmatic Access]: Filtered results information does not announce when user searches any File Name in Filter by Name field. [Fixed by #11512]
- [x] Bug 5: [Accessibility][Keyboard interfaces]: User is not able to select/unselect multiple file types using keyboard in the List. [Fixed by #7256]

<s>- Bug 6: [Accessibility][Programmatic access]: State is not announced properly  for the File which is selected in the List .</s>

- [x] Bug 7: [Accessibility][Programmatic access]: Name is not defined for 'Select all files' checkbox under 'File details list' which gets enable on activating "Enable toggle model selection"" button. [Fixed by #6682]
- [x] Bug 8: [Accessibility][Visual Requirements]: Tooltip is not provided for "File type " Icons. [Fixed by #6682]

<s>- Bug 9: [Accessibility][Screen Reader][Suggestion]: Scan mode focus should land on whole detail list row and it should announce all the information at the same time.</s>
<s>- Bug 10: [Functional]: 'Filter By' functionality  should be implemented for 'File Type 'list header columns.</s>

- [x] Bug 11: [Accessibility][Programmatic access]: 'Selected files count' is not being announced by the screen reader although selected files count appears on the screen. [Fixed by #11513]
- [x] Bug 12: [Accessibility][Programmatic access]: Screen reader speaks inappropriate information on 'File type column header'. [Fixed by #7256]

<s>- Bug 13: [Usable]: Information about the 'Details list navigation keys' is not provided when focus lands on First column header of the details list.</s>
<s>- Bug 14: [Accessibility][Usable]: Navigation and narration is not efficient through Details List in scan mode.</s>
<s>- Bug 15: [Accessibility][Usable]: Screen does not scroll down along with keyboard focus in back navigation.</s>
<s>- Bug 16: [Accessibility][Screen reader]: Narrator does not read notification alert properly in correct order.</s>
<s>- Bug 17: [Accessibility][Keyboard Navigation]: Tab Focus is moving out of the Notification window which appears on activating Create account' control in 'Command Bar Button' .</s>
<s>- Bug 18: [Accessibility][Keyboard Navigation]: Tab Focus is moving out of the Notification window which appears on activating 'Share w/split button'.</s>
<s>- Bug 19: [Accessibility][Screen reader]: Narrator does not read notification alert properly.</s>

 

Component: DetailsList
@micahgodbolt

**Comment by natalieethell** [comment]
https://github.com/OfficeDev/office-ui-fabric-react/pull/6682 should fix #7.

**Comment by natalieethell** [comment]
#8, too.

**Comment by natalieethell** [comment]
1 is a Toggle bug that is already being tracked here: https://github.com/OfficeDev/office-ui-fabric-react/issues/7134. I'll cross it out.

**Comment by natalieethell** [comment]
Bugs 15, 16, 17, 18, and 19 are already reported as bugs in other components elsewhere.
15: [Bug 6 in Button a11y pass](https://github.com/OfficeDev/office-ui-fabric-react/issues/6356)
16: [Bug 1 in Button a11y pass](https://github.com/OfficeDev/office-ui-fabric-react/issues/6356)
17: [Bug 3 in Button a11y pass](https://github.com/OfficeDev/office-ui-fabric-react/issues/6356)
18: [Bug 8 in Contextual Menu a11y pass](https://github.com/OfficeDev/office-ui-fabric-react/issues/6360)
19: [Bug 6 in Contextual Menu a11y pass](https://github.com/OfficeDev/office-ui-fabric-react/issues/6360)

I'll cross these out.

**Comment by TCambre** [comment]
Is there a status update on bug #3 -  Column sorting information is not announced by screen reader?

**Comment by natalieethell** [comment]
@TCambre from my understanding, it looks like we'll have to add an aria-live region to get the desired behavior. As Kevin mentioned above, the official ARIA example with aria-sort does not announce by screen reader either. 

The new Announced component we've been working on will be really helpful to use here. It's currently in experiments, but we're hoping to get that out to the office-ui-fabric-react package soon! (@betrue-final-final FYI)

**Comment by eveleens** [comment]
These two bugs may have to do with Bug 9, but I'm not sure. 

In Scan mode in DetailsList and hitting "down" arrow to scan elements, the user needs to click the down arrow twice to get past the checkmark - one to read the entire row's information, one to read the checkbox information.
**Actual:**
In scan mode navigation, Narrator Focus land on hidden elements and it reads unnecessary information. 
**Expected:**
In scan mode navigation, Narrator Focus should not land on hidden elements while navigating on the page. Focus should land on interactive elements and pass the information for the same control.

Then, we have another issue where arriving a row reads the entire row's information. 
**Actual Result:**
Screen reader pass all cell value information in one go.
**Expected result:**
Screen reader should not pass all cell value information in one go. Screen reader should read cell value with their respective column as per user action.

Both of these can be reproduced in the current version of the DetailsList. Let me know if you need more info!

**Comment by natalieethell** [comment]
Bug 3 should be fixed with this PR: https://github.com/OfficeDev/office-ui-fabric-react/pull/10433!

**Comment by khmakoto** [comment]
Bug 6 does not longer repro in current version, hence crossing it out.

**Comment by khmakoto** [comment]
Bug 9 is by design behavior of scan mode, since its inherent function is to break out of author's intent when it comes to navigation and dictation and go through everything in the DOM disregarding our focus stops. Given this, I'll be crossing out bug 9.

**Comment by khmakoto** [comment]
Bug 10 is a suggestion of a feature in the example and not something that is needed from an accessibility standpoint nor that we should bake into the component, hence crossing it out.

**Comment by xugao** [comment]
on Bug 13 - this is not a bug on the component but is related to our code example only. 
I don't think the aria label is needed since entering the detailslist announces the following:
>Table, has 501 rows, 6 columns, Toggle selection for all items check box unchecked, Column Header, ow 1 of 501 Column 1 of 6

This is enough information for user to know how to navigate further
